python/mspasspy/io/distributed.py

```diff
@@ -0,0 +1,1322 @@
+    :type exclude_keys: a :class:`list` of :class:`str`
+    :param format: Set the format of the parallel container to define the
+      dataset.   Must be either "spark" or "dask" or the job will abort
+      immediately with an exception
+    :type format: :class:`str`
+    :param spark_context: If using spark this argument is required.  Spark
+      defines the concept of a "context" that is a global control object that
+      manages schduling.  See online Spark documentation for details on
+      this concept.
+    :type spark_context: :class:`pyspark.SparkContext`
+    :param npartitions: The number of desired partitions for Dask or the number
+      of slices for Spark. By default Dask will use 100 and Spark will determine
+      it automatically based on the cluster.
+    :type npartitions: :class:`int`
+    :param data_tag:  The definition of a dataset can become ambiguous
+      when partially processed data are saved within a workflow.   A common
+      example would be windowing long time blocks of data to shorter time
+      windows around a particular seismic phase and saving the windowed data.
+      The windowed data can be difficult to distinguish from the original
+      with standard queries.  For this reason we make extensive use of "tags"
+      for save and read operations to improve the efficiency and simplify
+      read operations.   Default turns this off by setting the tag null (None).
+    :type data_tag: :class:`str`
+    :return: container defining the parallel dataset.  A spark `RDD` if format
+      is "Spark" and a dask 'bag' if format is "dask"
+    """
+    if not (isinstance(data, pd.DataFrame) or isinstance(data, Database)):
+        raise TypeError("Only Database or DataFrame are supported")
+    db = data
+    if isinstance(data, Database):
+        #  first read the metadata from database to a dataframe, and save to data
+        data = read_to_dataframe(
+            data,
+            cursor,
+            mode,
+            normalize,
+            load_history,
+            exclude_keys,
+            data_tag,
+        )
+
+    # now the type of data is a dataframe
+    if format == "spark":
+        list_ = spark_context.parallelize(
+            data.to_dict("records"), numSlices=npartitions
+        )
+    else:
+        list_ = daskbag.from_sequence(data.to_dict("records"), npartitions=npartitions)
+    # list_ is a parallel container of dict
+    return list_.map(
+        lambda cur: read_files(
+            Metadata(cur),  # convert dict to metadata
+            gridfs.GridFS(db)
+            if (isinstance(db, Database) and cur["storage_mode"] == "gridfs")
+            else None,
+            # if storage mode is gridfs, pass a GridFS object, it can not be put in the dataframe because rdd/bag
+            # can't pickle _thread.RLock objects
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key,
```

Comment:
These two arguments are missing the docstrings. I noticed it was missing from the old version, but we should add them.